Automattic/wp-calypso
# Allow Jetpack Stats checkout flows to use a redirect_to query arg.

Related to https://github.com/Automattic/red-team/issues/56 https://github.com/Automattic/jetpack-marketing/issues/1013

## Proposed Changes

* Support `redirect_to` for the Jetpack siteless checkout flow
* Detect site connection in Jetpack Stats and pass the correct value for `connect_after_checkout`; otherwise, the Stats checkout would be broken for the updated flow.

## Why are these changes being made?

* Allow user to return to where they initialized the purchase

## Testing Instructions

(1) Prerequisite:
You need a jetpack site that is either _not site or user connected_, -or- _only site connected and not user connected_.
Make sure you are logged into wordpress.com with your A11n account so you can purchase with credits.

* Checkout this branch and `yarn-start`.
* Go to your jetpack site, My Jetpack.  Ensure your site is either not user connected, or not site and not user connected.
* In My Jetpack, click "Get Plan" on the Backup product card.
* Click "Get Jetpack VaultPress Backup", you'll be taken to calypso checkout
* On checkout, in the URL, replace `https://wordpress.com` with `http://calypso.localhost:3000`
* Also in the url change the `&redirect_to=` query arg value to something like, `https://yoursitedomain.com/wp-admin/admin.php?page=testing_this_redirect`, just so you'll definitely know it works when you see the final redirect location.
* Go ahead and complete checkout (using credits is fine).
* After checkout is completed, your site should auto connect and you should first be redirected to the connection authorization page.  Click "Authorize", then you should be redirectd to the product licensing thank-you auto activation page. The product license should auto-activate.
* Finally once all complete, you should see the "Go to Dashboard" button. Inspect the button and see that the href of the button is the `redirect_to` url you update in the step above.

(2) Ensure you can purchase Stats from connected Jetpack self hosted site 

- Prepare a Jetpack site that is fully connected, and build Jetpack Stats locally following PCYsg-Pp7-p2
- Open `/wp-admin/admin.php?page=stats#!/stats/purchase/:siteSlug?from=jetpack-stats-tier-upgrade-usage-section&productType=commercial`
- Click 'Upgrade now and continue'
- Replace `https://wordpress.com` with `http://calypso.localhost:3000`
- Ensure you are able to purchase Stats via siteless flow - buy first and then assign the purchased license

(3) Ensure you can purchase Stats from not fully connected Jetpack self hosted site

- Prepare a Jetpack site that is NOT fully connected, repeat the steps in (2)


(4) Ensure you can purchase Stats from Calypso Stats 
- Prepare a Jetpack site that is fully connected
- Manually open a Jetpack site in Calypso, example: `http://calypso.localhost:3000/stats/day/:siteSlug`
- Open `http://calypso.localhost:3000/stats/purchase/:siteSlug?from=jetpack-stats-tier-upgrade-usage-section&productType=commercial`
- Ensure you are able to purchase Stats with no issues


![image](https://github.com/user-attachments/assets/ee1a6a8f-9f46-4180-9ac4-6f9824f67bb9)


## Pre-merge Checklist
